services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java

```diff
@@ -0,0 +1,352 @@
+   */
+  @Override
+  public void notifyOnBuild(Server server) {
+    if (this.server == null) {
+      this.server = Preconditions.checkNotNull(server, "server");
+    }
+  }
+
+  private void processExtension(FieldDescriptor extension, FileDescriptor fd) {
+    String extensionName = extension.getContainingType().getFullName();
+    int extensionNumber = extension.getNumber();
+    if (!fileDescriptorsByExtensionAndNumber.containsKey(extensionName)) {
+      fileDescriptorsByExtensionAndNumber.put(extensionName,
+          new HashMap<Integer, FileDescriptor>());
+    }
+    Preconditions.checkState(
+        !fileDescriptorsByExtensionAndNumber.get(extensionName).containsKey(extensionNumber),
+        "Extension name and number already defined: %s, %s", extensionName, extensionNumber);
+    fileDescriptorsByExtensionAndNumber.get(extensionName).put(extensionNumber, fd);
+  }
+
+  private void processService(ServiceDescriptor service, FileDescriptor fd) {
+    String serviceName = service.getFullName();
+    Preconditions.checkState(!fileDescriptorsBySymbol.containsKey(serviceName),
+        "Service already defined: %s", serviceName);
+    fileDescriptorsBySymbol.put(serviceName, fd);
+    for (MethodDescriptor method : service.getMethods()) {
+      String methodName = method.getFullName();
+      Preconditions.checkState(!fileDescriptorsBySymbol.containsKey(methodName),
+          "Method already defined: %s", methodName);
+      fileDescriptorsBySymbol.put(methodName, fd);
+    }
+  }
+
+  private void processType(Descriptor type, FileDescriptor fd) {
+    String typeName = type.getFullName();
+    Preconditions.checkState(!fileDescriptorsBySymbol.containsKey(typeName),
+        "Type already defined: %s", typeName);
+    fileDescriptorsBySymbol.put(typeName, fd);
+    for (FieldDescriptor extension : type.getExtensions()) {
+      processExtension(extension, fd);
+    }
+    for (Descriptor nestedType : type.getNestedTypes()) {
+      processType(nestedType, fd);
+    }
+  }
+
+  private void processFileDescriptor(FileDescriptor fd) {
+    String fdName = fd.getName();
+    Preconditions.checkState(!fileDescriptorsByName.containsKey(fdName),
+        "File name already used: %s", fdName);
+    fileDescriptorsByName.put(fdName, fd);
+    for (ServiceDescriptor service : fd.getServices()) {
+      processService(service, fd);
+    }
+    for (Descriptor type : fd.getMessageTypes()) {
+      processType(type, fd);
+    }
+    for (FieldDescriptor extension : fd.getExtensions()) {
+      processExtension(extension, fd);
+    }
+  }
+
+  /**
+   * Processes the service file descriptors. This prepares a list of the service names and indexes
+   * the file descriptors by name, symbol, and extension.
+   */
+  private synchronized void initFileDescriptorMaps() {
+    if (mapsInitialized) {
+      // Check if the server's set of services has changed
+      List<ServerServiceDefinition> currentServices = server.getServices();
```

Comment:
In earlier discussions with @ejona86 we talked about supporting dynamic services. I dropped this check and just re-process all file descriptors for each query, which still supports dynamic services but removes some of the complexity of initialization.